File: ubag-node/src/provenance.js
'use strict';

/**
 * UBAG Provenance Profile 1.0 draft.
 *
 * This module ends at VerifiedProvenance. Tenant/principal resolution belongs
 * to the private gateway and must never consume client-authored identity claims.
 */
const crypto = require('crypto');
const { MemoryReplayStore, ReplayStoreCapacityError } = require('./challenge');
const { agentSign, agentVerify } = require('./keys');

const PROFILE_TYPE = 'ubag-provenance+jwt';
const MAX_LIFETIME = 60;
const MAX_CLOCK_SKEW = 5;
const MAX_JTI_BYTES = 128;
const MAX_ENVELOPE_BYTES = 65536;
const MAX_HEADER_BYTES = 4096;
const MAX_PAYLOAD_BYTES = 16384;
const HEADER_FIELDS = new Set(['typ', 'alg', 'jwk', 'kid']);
const PAYLOAD_FIELDS = new Set([
  'jti', 'htm', 'htu', 'iat', 'exp', 'ath', 'ubag_qsh',
  'ubag_content_digest', 'nonce',
]);
const FORBIDDEN_FIELDS = new Set(['tenant', 'tenant_ref', 'principal', 'principal_ref']);
// Replay identifiers are "{originHash}:{agentHash}|{jti}". Both hashes are
// unpadded base64url, so neither can contain ':' or '|', while a client-controlled
// jti may contain both. Taking element [0] is what makes that safe, and it is also
// what keeps the two SDKs identical: JavaScript's split(sep, limit) bounds the
// number of returned elements and Python's split(sep, maxsplit) bounds the number
// of splits, so only [0] agrees under both. Do not replace either extractor with
// lastIndexOf-based parsing or a regex.
const defaultProvenanceStore = new MemoryReplayStore(100000, {
  maxEntriesPerNamespace: 4096,
  namespaceExtractor: (identifier) => identifier.split('|', 1)[0],
  maxEntriesPerGroup: 20000,
  groupExtractor: (identifier) => identifier.split(':', 1)[0],
});

class ProvenanceError extends Error {
  constructor(code, message = '', { agentRef = null } = {}) {
    super(message || code);
    this.code = code;
    this.agentRef = agentRef;
  }
}

const b64u = (value) => Buffer.from(value).toString('base64url');
const decode = (value) => {
  if (typeof value !== 'string' || value.length === 0) throw new Error('invalid base64url');
  return Buffer.from(value, 'base64url');
};

function parseProvenance(compactJws) {
  try {
    if (typeof compactJws !== 'string') throw new Error('string required');
    if (Buffer.byteLength(compactJws, 'utf8') > MAX_ENVELOPE_BYTES) {
      throw new Error('envelope too large');
    }
    const parts = compactJws.split('.');
    if (parts.length !== 3 || parts.some((part) => !part)) throw new Error('three parts required');
    const headerBytes = decode(parts[0]);
    const payloadBytes = decode(parts[1]);
    if (headerBytes.length > MAX_HEADER_BYTES || payloadBytes.length > MAX_PAYLOAD_BYTES) {
      throw new Error('JWS component too large');
    }
    const header = JSON.parse(headerBytes.toString('utf8'));
    const payload = JSON.parse(payloadBytes.toString('utf8'));
    if (!header || Array.isArray(header) || typeof header !== 'object' ||
        !payload || Array.isArray(payload) || typeof payload !== 'object') {
      throw new Error('objects required');
    }
    validateNumberDomain(header);
    validateNumberDomain(payload);
    if (!headerBytes.equals(Buffer.from(JSON.stringify(stable(header)), 'utf8'))) {
      throw new Error('protected header is not RFC 8785 canonical JSON');
    }
    if (!payloadBytes.equals(Buffer.from(JSON.stringify(stable(payload)), 'utf8'))) {
      throw new Error('payload is not RFC 8785 canonical JSON');
    }
    if (Object.keys(header).some((key) => !HEADER_FIELDS.has(key)) || header.crit !== undefined) {
      throw new ProvenanceError('UNSUPPORTED_PROFILE');
    }
    if (Object.keys(payload).some((key) => FORBIDDEN_FIELDS.has(key))) {
      throw new ProvenanceError('INVALID_CLAIMS');
    }
    if (Object.keys(payload).some((key) => !PAYLOAD_FIELDS.has(key))) {
      throw new ProvenanceError('INVALID_CLAIMS');
    }
    return {
      compactJws,
      header,
      payload,
      payloadBytes,
      signingInput: Buffer.from(`${parts[0]}.${parts[1]}`, 'ascii'),
      signature: parts[2],
    };
  } catch (error) {
    if (error instanceof ProvenanceError) throw error;
    throw new ProvenanceError('MALFORMED_ENVELOPE');
  }
}

function contentDigest(body = Buffer.alloc(0)) {
  const raw = Buffer.isBuffer(body) ? body : Buffer.from(body);
  return `sha-256=:${crypto.createHash('sha256').update(raw).digest('base64')}:`;
}

function validateNumberDomain(value) {
  if (value === null || typeof value === 'string' || typeof value === 'boolean') return;
  if (typeof value === 'number') {
    if (!Number.isSafeInteger(value)) {
      throw new ProvenanceError('MALFORMED_ENVELOPE',
        'numbers must be interoperable safe integers');
    }
    return;
  }
  if (Array.isArray(value)) {
    value.forEach(validateNumberDomain);
    return;
  }
  if (value && typeof value === 'object') {
    Object.values(value).forEach(validateNumberDomain);
    return;
  }
  throw new ProvenanceError('MALFORMED_ENVELOPE');
}

function queryStringHash(originForm) {
  if (typeof originForm !== 'string' || !originForm.startsWith('/') || originForm.includes('#')) {
    throw new Error("originForm must begin with '/' and contain no fragment");
  }
  return crypto.createHash('sha256').update(Buffer.from(originForm, 'utf8')).digest('base64url');
}

function jwkThumbprint(publicKey) {
  const canonical = JSON.stringify({ crv: 'Ed25519', kty: 'OKP', x: publicKey });
  return crypto.createHash('sha256').update(canonical).digest('base64url');
}

function agentRef(publicKey) {
  return `ubag:sha256:${crypto.createHash('sha256').update(decode(publicKey)).digest('base64url')}`;
}

function credentialHash(token) {
  return crypto.createHash('sha256').update(String(token), 'utf8').digest('base64url');
}

function checkedPublicKey(header) {
  if (header.typ !== PROFILE_TYPE || header.alg !== 'EdDSA') {
    throw new ProvenanceError('UNSUPPORTED_PROFILE');
  }
  const jwk = header.jwk;
  if (!jwk || Array.isArray(jwk) || typeof jwk !== 'object' ||
      Object.keys(jwk).sort().join(',') !== 'crv,kty,x' ||
      jwk.kty !== 'OKP' || jwk.crv !== 'Ed25519') {
    throw new ProvenanceError('UNSUPPORTED_PROFILE');
  }
  if (typeof jwk.x !== 'string' || decode(jwk.x).length !== 32) {
    throw new ProvenanceError('UNSUPPORTED_PROFILE');
  }
  return jwk.x;
}

function verifyProvenance(compactJws, {
  method,
  publicOrigin,
  originForm,
  body = Buffer.alloc(0),
  credential = null,
  credentialValidator = null,
  replayStore = null,
  expectedNonce = null,
  now = null,
}) {
  const envelope = parseProvenance(compactJws);
  const publicKey = checkedPublicKey(envelope.header);
  if (!agentVerify(publicKey, envelope.signingInput, envelope.signature)) {
    throw new ProvenanceError('INVALID_SIGNATURE');
  }
  const derivedAgentRef = agentRef(publicKey);
  if (envelope.header.kid !== undefined && envelope.header.kid !== derivedAgentRef) {
    throw new ProvenanceError('KEY_HINT_MISMATCH');
  }
  const p = envelope.payload;
  const required = ['jti', 'htm', 'htu', 'iat', 'exp', 'ubag_qsh', 'ubag_content_digest'];
  if (required.some((name) => p[name] === undefined) ||
      ['jti', 'htm', 'htu', 'ubag_qsh', 'ubag_content_digest']
        .some((name) => typeof p[name] !== 'string' || p[name].length === 0) ||
      !Number.isInteger(p.iat) || !Number.isInteger(p.exp) ||
      p.exp <= p.iat || p.exp - p.iat > MAX_LIFETIME) {
    throw new ProvenanceError('INVALID_CLAIMS');
  }
  if (Buffer.byteLength(p.jti, 'utf8') > MAX_JTI_BYTES) {
    throw new ProvenanceError('INVALID_CLAIMS', 'jti exceeds 128 bytes');
  }
  const current = now === null ? Math.floor(Date.now() / 1000) : Number(now);
  if (p.iat > current + MAX_CLOCK_SKEW || p.exp < current - MAX_CLOCK_SKEW) {
    throw new ProvenanceError('PROOF_EXPIRED');
  }
  const expectedMethod = String(method).toUpperCase();
  if (p.htm !== expectedMethod) throw new ProvenanceError('METHOD_MISMATCH');
  const path = originForm.split('?', 1)[0];
  const expectedHtu = String(publicOrigin).replace(/\/+$/, '') + path;
  if (p.htu !== expectedHtu) throw new ProvenanceError('ORIGIN_MISMATCH');
  if (p.ubag_qsh !== queryStringHash(originForm)) throw new ProvenanceError('TARGET_MISMATCH');
  const expectedDigest = contentDigest(body);
  if (p.ubag_content_digest !== expectedDigest) {
    throw new ProvenanceError('CONTENT_DIGEST_MISMATCH');
  }

  let claims = null;
  if (credential === null || credential === undefined) {
    if (Object.prototype.hasOwnProperty.call(p, 'ath')) {
      throw new ProvenanceError('CREDENTIAL_MISMATCH');
    }
  } else {
    if (typeof credentialValidator !== 'function') {
      throw new ProvenanceError('VERIFIER_MISCONFIGURED',
        'credential validator is required');
    }
    if (p.ath !== credentialHash(credential)) {
      throw new ProvenanceError('CREDENTIAL_MISMATCH');
    }
    claims = credentialValidator(credential);
    if (!claims || typeof claims !== 'object' || !claims.cnf ||
        claims.cnf.jkt !== jwkThumbprint(publicKey)) {
      throw new ProvenanceError('CREDENTIAL_MISMATCH');
    }
  }

  if (expectedNonce !== null) {
    if (p.nonce !== expectedNonce) throw new ProvenanceError('NONCE_MISMATCH');
  } else if (p.nonce !== undefined) {
    throw new ProvenanceError('NONCE_MISMATCH');
  }

  const store = replayStore || defaultProvenanceStore;
  const originNamespace = crypto.createHash('sha256')
    .update(String(publicOrigin).replace(/\/+$/, ''), 'utf8')
    .digest('base64url');
  const agentNamespace = crypto.createHash('sha256')
    .update(derivedAgentRef, 'utf8')
    .digest('base64url');
  let consumed;
  try {
    consumed = store.consume(`${originNamespace}:${agentNamespace}|${p.jti}`,
      p.exp + MAX_CLOCK_SKEW + 1);
  } catch (error) {
    if (error instanceof ReplayStoreCapacityError) {
      if (error.tier === 'namespace') {
        throw new ProvenanceError('REPLAY_KEY_QUOTA_EXCEEDED', '', {
          agentRef: derivedAgentRef,
        });
      }
      if (error.tier === 'group') {
        throw new ProvenanceError('REPLAY_ORIGIN_QUOTA_EXCEEDED');
      }
      throw new ProvenanceError('REPLAY_STORE_EXHAUSTED');
    }
    throw error;
  }
  if (!consumed) {
    throw new ProvenanceError('PROVENANCE_REPLAYED');
  }
  return Object.freeze({
    agentRef: derivedAgentRef,
    publicKey,
    credentialRef: claims ? claims.jti : null,
    issuer: claims ? claims.iss : null,
    audience: claims ? claims.aud : null,
    method: expectedMethod,
    htu: expectedHtu,
    originForm,
    contentDigest: expectedDigest,
    jti: p.jti,
    issuedAt: p.iat,
    expiresAt: p.exp,
    nonce: p.nonce || null,
  });
}

function stable(value) {
  if (Array.isArray(value)) return value.map(stable);
  if (value && typeof value === 'object') {
    return Object.fromEntries(Object.keys(value).sort().map((key) => [key, stable(value[key])]));
  }
  return value;
}

function createProvenance(privateKey, publicKey, {
  method,
  publicOrigin,
  originForm,
  body = Buffer.alloc(0),
  credential = null,
  jti,
  issuedAt = Math.floor(Date.now() / 1000),
  lifetime = MAX_LIFETIME,
  nonce = null,
  kid = null,
}) {
  if (!Number.isInteger(lifetime) || lifetime < 1 || lifetime > MAX_LIFETIME) {
    throw new Error('lifetime must be between 1 and 60 seconds');
  }
  const path = originForm.split('?', 1)[0];
  const header = {
    typ: PROFILE_TYPE,
    alg: 'EdDSA',
    jwk: { kty: 'OKP', crv: 'Ed25519', x: publicKey },
  };
  if (kid) {
    if (kid !== agentRef(publicKey)) throw new Error('kid must equal the reference derived from publicKey');
    header.kid = kid;
  }
  const payload = {
    jti,
    htm: String(method).toUpperCase(),
    htu: String(publicOrigin).replace(/\/+$/, '') + path,
    iat: Number(issuedAt),
    exp: Number(issuedAt) + lifetime,
    ubag_qsh: queryStringHash(originForm),
    ubag_content_digest: contentDigest(body),
  };
  if (credential !== null && credential !== undefined) payload.ath = credentialHash(credential);
  if (nonce !== null && nonce !== undefined) payload.nonce = nonce;
  const encodedHeader = b64u(Buffer.from(JSON.stringify(stable(header))));
  const encodedPayload = b64u(Buffer.from(JSON.stringify(stable(payload))));
  const signingInput = `${encodedHeader}.${encodedPayload}`;
  return `${signingInput}.${agentSign(privateKey, signingInput)}`;
}

module.exports = {
  PROFILE_TYPE,
  ProvenanceError,
  parseProvenance,
  contentDigest,
  queryStringHash,
  jwkThumbprint,
  agentRef,
  credentialHash,
  verifyProvenance,
  createProvenance,
};

File: ubag-python/src/ubag/provenance.py
"""UBAG Provenance Profile 1.0 draft.

The wire object is untrusted until :func:`verify_provenance` returns a
``VerifiedProvenance``. Tenant and principal deliberately do not exist in this
module: they belong to the private resolver.
"""
from __future__ import annotations

import base64
import hashlib
import json
import time
from dataclasses import dataclass
from typing import Callable, Optional

from cryptography.hazmat.primitives.asymmetric.ed25519 import (
    Ed25519PrivateKey,
    Ed25519PublicKey,
)

from ubag._challenge import (
    MemoryReplayStore,
    ReplayStoreCapacityError,
    UBAGReplayStore,
)

PROFILE_TYPE = "ubag-provenance+jwt"
MAX_LIFETIME = 60
MAX_CLOCK_SKEW = 5
MAX_JTI_BYTES = 128
MAX_ENVELOPE_BYTES = 65_536
MAX_HEADER_BYTES = 4_096
MAX_PAYLOAD_BYTES = 16_384
MAX_SAFE_INTEGER = 9_007_199_254_740_991
_HEADER_FIELDS = {"typ", "alg", "jwk", "kid"}
_PAYLOAD_FIELDS = {
    "jti", "htm", "htu", "iat", "exp", "ath", "ubag_qsh",
    "ubag_content_digest", "nonce",
}
_FORBIDDEN_FIELDS = {"tenant", "tenant_ref", "principal", "principal_ref"}
# Replay identifiers are "{origin_hash}:{agent_hash}|{jti}". Both hashes are
# unpadded base64url, so neither can contain ":" or "|", while a client-controlled
# jti may contain both. Taking element [0] is what makes that safe, and it is also
# what keeps the two SDKs identical: Python's split(sep, maxsplit) bounds the
# number of splits and JavaScript's split(sep, limit) bounds the number of
# returned elements, so only [0] agrees under both. Do not replace either
# extractor with rsplit or a regex.
_default_provenance_store = MemoryReplayStore(
    max_entries=100_000,
    max_entries_per_namespace=4_096,
    namespace_extractor=lambda identifier: identifier.split("|", 1)[0],
    max_entries_per_group=20_000,
    group_extractor=lambda identifier: identifier.split(":", 1)[0],
)


class ProvenanceError(ValueError):
    def __init__(
        self,
        code: str,
        message: str = "",
        *,
        agent_ref: Optional[str] = None,
    ) -> None:
        super().__init__(message or code)
        self.code = code
        self.agent_ref = agent_ref


@dataclass(frozen=True)
class ProvenanceEnvelope:
    compact_jws: str
    header: dict
    payload: dict
    payload_bytes: bytes
    signing_input: bytes
    signature: bytes


@dataclass(frozen=True)
class VerifiedProvenance:
    agent_ref: str
    public_key: str
    credential_ref: Optional[str]
    issuer: Optional[str]
    audience: Optional[object]
    method: str
    htu: str
    origin_form: str
    content_digest: str
    jti: str
    issued_at: int
    expires_at: int
    nonce: Optional[str] = None


def _b64u(data: bytes) -> str:
    return base64.urlsafe_b64encode(data).decode("ascii").rstrip("=")


def _b64u_decode(value: str) -> bytes:
    if not isinstance(value, str) or not value:
        raise ValueError("invalid base64url")
    return base64.urlsafe_b64decode(value + "=" * (-len(value) % 4))


def _strict_json(data: bytes) -> dict:
    def pairs(values):
        result = {}
        for key, value in values:
            if key in result:
                raise ValueError("duplicate JSON member")
            result[key] = value
        return result

    def bad_constant(value):
        raise ValueError(f"non-I-JSON number: {value}")

    value = json.loads(
        data.decode("utf-8", errors="strict"),
        object_pairs_hook=pairs,
        parse_constant=bad_constant,
    )
    if not isinstance(value, dict):
        raise ValueError("JSON object required")
    return value


def _jcs(value: dict) -> bytes:
    # Profile claims contain strings, integers, and flat JSON objects only; this
    # is RFC 8785 canonical form for that deliberately bounded value domain.
    return json.dumps(
        value, ensure_ascii=False, separators=(",", ":"), sort_keys=True
    ).encode("utf-8")


def _validate_json_number_domain(value) -> None:
    if isinstance(value, bool) or value is None or isinstance(value, str):
        return
    if isinstance(value, float):
        raise ProvenanceError("MALFORMED_ENVELOPE",
                              "floating-point claims are outside the profile")
    if isinstance(value, int):
        if abs(value) > MAX_SAFE_INTEGER:
            raise ProvenanceError("MALFORMED_ENVELOPE",
                                  "integer exceeds interoperable safe range")
        return
    if isinstance(value, list):
        for item in value:
            _validate_json_number_domain(item)
        return
    if isinstance(value, dict):
        for item in value.values():
            _validate_json_number_domain(item)
        return
    raise ProvenanceError("MALFORMED_ENVELOPE")


def parse_provenance(compact_jws: str) -> ProvenanceEnvelope:
    try:
        if not isinstance(compact_jws, str):
            raise ValueError
        if len(compact_jws.encode("utf-8")) > MAX_ENVELOPE_BYTES:
            raise ValueError("envelope too large")
        parts = compact_jws.split(".")
        if len(parts) != 3 or not all(parts):
            raise ValueError
        header_raw, payload_raw, signature_raw = map(_b64u_decode, parts)
        if len(header_raw) > MAX_HEADER_BYTES or len(payload_raw) > MAX_PAYLOAD_BYTES:
            raise ValueError("JWS component too large")
        header = _strict_json(header_raw)
        payload = _strict_json(payload_raw)
        _validate_json_number_domain(header)
        _validate_json_number_domain(payload)
        if header_raw != _jcs(header):
            raise ValueError("protected header is not RFC 8785 canonical JSON")
        if payload_raw != _jcs(payload):
            raise ValueError("payload is not RFC 8785 canonical JSON")
        signature = signature_raw
        if set(header) - _HEADER_FIELDS or "crit" in header:
            raise ProvenanceError("UNSUPPORTED_PROFILE", "unsupported protected header")
        if set(payload) & _FORBIDDEN_FIELDS:
            raise ProvenanceError("INVALID_CLAIMS", "identity resolution claim on wire")
        if set(payload) - _PAYLOAD_FIELDS:
            raise ProvenanceError("INVALID_CLAIMS", "unknown payload claim")
        return ProvenanceEnvelope(
            compact_jws=compact_jws,
            header=header,
            payload=payload,
            payload_bytes=payload_raw,
            signing_input=f"{parts[0]}.{parts[1]}".encode("ascii"),
            signature=signature,
        )
    except ProvenanceError:
        raise
    except Exception as exc:
        raise ProvenanceError("MALFORMED_ENVELOPE") from exc


def content_digest(body: bytes) -> str:
    """RFC 9530 sha-256 dictionary member for the exact content bytes."""
    if not isinstance(body, bytes):
        raise TypeError("body must be bytes")
    digest = base64.b64encode(hashlib.sha256(body).digest()).decode("ascii")
    return f"sha-256=:{digest}:"


def query_string_hash(origin_form: str) -> str:
    """Hash exact origin-form UTF-8 bytes without parsing or normalization."""
    if not isinstance(origin_form, str) or not origin_form.startswith("/"):
        raise ValueError("origin_form must begin with '/'")
    if "#" in origin_form:
        raise ValueError("origin_form must not contain a fragment")
    return _b64u(hashlib.sha256(origin_form.encode("utf-8")).digest())


def jwk_thumbprint(public_b64: str) -> str:
    canonical = (
        '{"crv":"Ed25519","kty":"OKP","x":' + json.dumps(public_b64, separators=(",", ":")) + "}"
    ).encode("utf-8")
    return _b64u(hashlib.sha256(canonical).digest())


def agent_ref(public_b64: str) -> str:
    return "ubag:sha256:" + _b64u(hashlib.sha256(_b64u_decode(public_b64)).digest())


def credential_hash(token: str) -> str:
    return _b64u(hashlib.sha256(token.encode("utf-8")).digest())


def _public_key(header: dict) -> tuple[str, Ed25519PublicKey]:
    if header.get("typ") != PROFILE_TYPE or header.get("alg") != "EdDSA":
        raise ProvenanceError("UNSUPPORTED_PROFILE")
    jwk = header.get("jwk")
    if not isinstance(jwk, dict) or set(jwk) != {"kty", "crv", "x"}:
        raise ProvenanceError("UNSUPPORTED_PROFILE", "exact Ed25519 public JWK required")
    if jwk.get("kty") != "OKP" or jwk.get("crv") != "Ed25519":
        raise ProvenanceError("UNSUPPORTED_PROFILE")
    try:
        public_raw = _b64u_decode(jwk["x"])
        if len(public_raw) != 32:
            raise ValueError
        return jwk["x"], Ed25519PublicKey.from_public_bytes(public_raw)
    except Exception as exc:
        raise ProvenanceError("UNSUPPORTED_PROFILE", "invalid Ed25519 JWK") from exc


def verify_provenance(
    compact_jws: str,
    *,
    method: str,
    public_origin: str,
    origin_form: str,
    body: bytes = b"",
    credential: Optional[str] = None,
    credential_validator: Optional[Callable[[str], Optional[dict]]] = None,
    replay_store: Optional[UBAGReplayStore] = None,
    expected_nonce: Optional[str] = None,
    now: Optional[int] = None,
) -> VerifiedProvenance:
    envelope = parse_provenance(compact_jws)
    public_b64, key = _public_key(envelope.header)
    try:
        key.verify(envelope.signature, envelope.signing_input)
    except Exception as exc:
        raise ProvenanceError("INVALID_SIGNATURE") from exc
    derived_agent_ref = agent_ref(public_b64)
    if "kid" in envelope.header and envelope.header["kid"] != derived_agent_ref:
        raise ProvenanceError("KEY_HINT_MISMATCH")

    p = envelope.payload
    required = {"jti", "htm", "htu", "iat", "exp", "ubag_qsh", "ubag_content_digest"}
    if not required.issubset(p):
        raise ProvenanceError("INVALID_CLAIMS", "required claim missing")
    if not all(isinstance(p.get(name), str) and p[name] for name in
               ("jti", "htm", "htu", "ubag_qsh", "ubag_content_digest")):
        raise ProvenanceError("INVALID_CLAIMS")
    if len(p["jti"].encode("utf-8")) > MAX_JTI_BYTES:
        raise ProvenanceError("INVALID_CLAIMS", "jti exceeds 128 bytes")
    if isinstance(p.get("iat"), bool) or isinstance(p.get("exp"), bool):
        raise ProvenanceError("INVALID_CLAIMS")
    if not isinstance(p.get("iat"), int) or not isinstance(p.get("exp"), int):
        raise ProvenanceError("INVALID_CLAIMS")
    if p["exp"] <= p["iat"] or p["exp"] - p["iat"] > MAX_LIFETIME:
        raise ProvenanceError("INVALID_CLAIMS", "invalid proof lifetime")
    current = int(time.time()) if now is None else int(now)
    if p["iat"] > current + MAX_CLOCK_SKEW or p["exp"] < current - MAX_CLOCK_SKEW:
        raise ProvenanceError("PROOF_EXPIRED")

    expected_method = method.upper()
    if p["htm"] != expected_method:
        raise ProvenanceError("METHOD_MISMATCH")
    path = origin_form.split("?", 1)[0]
    expected_htu = public_origin.rstrip("/") + path
    if p["htu"] != expected_htu:
        raise ProvenanceError("ORIGIN_MISMATCH")
    if p["ubag_qsh"] != query_string_hash(origin_form):
        raise ProvenanceError("TARGET_MISMATCH")
    expected_digest = content_digest(body)
    if p["ubag_content_digest"] != expected_digest:
        raise ProvenanceError("CONTENT_DIGEST_MISMATCH")

    claims = None
    if credential is None:
        if "ath" in p:
            raise ProvenanceError("CREDENTIAL_MISMATCH")
    else:
        if credential_validator is None:
            raise ProvenanceError("VERIFIER_MISCONFIGURED",
                                  "credential validator is required")
        if p.get("ath") != credential_hash(credential):
            raise ProvenanceError("CREDENTIAL_MISMATCH")
        claims = credential_validator(credential)
        if not isinstance(claims, dict):
            raise ProvenanceError("CREDENTIAL_MISMATCH")
        cnf = claims.get("cnf")
        if not isinstance(cnf, dict) or cnf.get("jkt") != jwk_thumbprint(public_b64):
            raise ProvenanceError("CREDENTIAL_MISMATCH")

    supplied_nonce = p.get("nonce")
    if expected_nonce is not None:
        if supplied_nonce != expected_nonce:
            raise ProvenanceError("NONCE_MISMATCH")
    elif supplied_nonce is not None:
        raise ProvenanceError("NONCE_MISMATCH", "unsolicited nonce")

    store = replay_store if replay_store is not None else _default_provenance_store
    origin_namespace = _b64u(hashlib.sha256(
        public_origin.rstrip("/").encode("utf-8")
    ).digest())
    agent_namespace = _b64u(hashlib.sha256(
        derived_agent_ref.encode("utf-8")
    ).digest())
    try:
        consumed = store.consume(
            f"{origin_namespace}:{agent_namespace}|{p['jti']}",
            p["exp"] + MAX_CLOCK_SKEW + 1,
        )
    except ReplayStoreCapacityError as exc:
        if exc.tier == "namespace":
            raise ProvenanceError(
                "REPLAY_KEY_QUOTA_EXCEEDED",
                agent_ref=derived_agent_ref,
            ) from exc
        if exc.tier == "group":
            raise ProvenanceError("REPLAY_ORIGIN_QUOTA_EXCEEDED") from exc
        raise ProvenanceError("REPLAY_STORE_EXHAUSTED") from exc
    if not consumed:
        raise ProvenanceError("PROVENANCE_REPLAYED")

    return VerifiedProvenance(
        agent_ref=derived_agent_ref,
        public_key=public_b64,
        credential_ref=claims.get("jti") if claims else None,
        issuer=claims.get("iss") if claims else None,
        audience=claims.get("aud") if claims else None,
        method=expected_method,
        htu=expected_htu,
        origin_form=origin_form,
        content_digest=expected_digest,
        jti=p["jti"],
        issued_at=p["iat"],
        expires_at=p["exp"],
        nonce=supplied_nonce,
    )


def create_provenance(
    private_b64: str,
    public_b64: str,
    *,
    method: str,
    public_origin: str,
    origin_form: str,
    body: bytes = b"",
    credential: Optional[str] = None,
    jti: str,
    issued_at: Optional[int] = None,
    lifetime: int = MAX_LIFETIME,
    nonce: Optional[str] = None,
    kid: Optional[str] = None,
) -> str:
    if not 1 <= lifetime <= MAX_LIFETIME:
        raise ValueError("lifetime must be between 1 and 60 seconds")
    issued = int(time.time()) if issued_at is None else int(issued_at)
    path = origin_form.split("?", 1)[0]
    header = {
        "typ": PROFILE_TYPE,
        "alg": "EdDSA",
        "jwk": {"kty": "OKP", "crv": "Ed25519", "x": public_b64},
    }
    if kid:
        if kid != agent_ref(public_b64):
            raise ValueError("kid must equal the reference derived from public_b64")
        header["kid"] = kid
    payload = {
        "jti": jti,
        "htm": method.upper(),
        "htu": public_origin.rstrip("/") + path,
        "iat": issued,
        "exp": issued + lifetime,
        "ubag_qsh": query_string_hash(origin_form),
        "ubag_content_digest": content_digest(body),
    }
    if credential is not None:
        payload["ath"] = credential_hash(credential)
    if nonce is not None:
        payload["nonce"] = nonce
    encoded_header = _b64u(_jcs(header))
    encoded_payload = _b64u(_jcs(payload))
    signing_input = f"{encoded_header}.{encoded_payload}".encode("ascii")
    private = Ed25519PrivateKey.from_private_bytes(_b64u_decode(private_b64))
    return f"{encoded_header}.{encoded_payload}.{_b64u(private.sign(signing_input))}"
